src/web.rs

```diff
@@ -328,37 +339,75 @@ async fn get_ui(
+                    }
+                    Err(e) => {
+                        tracing::error!("Failed to read file {:?}: {e}", &file_path);
+                        (
+                            StatusCode::INTERNAL_SERVER_ERROR,
+                            ([(header::CONTENT_TYPE, "text/html")]),
+                            "<body><h1>500 Internal Server Error</h1><p>Failed to read file in override web-ui directory</p></body>",
+                        )
+                            .into_response()
+                    }
+                }
+            } else {
+                (
+                    StatusCode::NOT_FOUND,
+                    ([(header::CONTENT_TYPE, "text/html")]),
+                    "<body><h1>404 Not Found</h1></body>",
+                )
+                    .into_response()
+            }
+        }
+        None => {
+            state.ui.map_or_else(
                 || {
                     (
-                        StatusCode::NOT_FOUND,
-                        ([(header::CONTENT_TYPE, "text/html")]),
-                        "<body><h1>404 Not Found</h1></body>",
-                    )
-                        .into_response()
+                    StatusCode::NOT_FOUND,
+                    ([(header::CONTENT_TYPE, "text/html")]),
+                    "<body><h1>There is no UI bundled with this version of the application.</h1></body>",
+                )
+                    .into_response()
                 },
-                |file| {
-                    // Serve included file
-                    let content_type = guess_content_type(file.path());
-                    let headers = [
-                        (header::CONTENT_TYPE, content_type),
-                        (header::ACCESS_CONTROL_ALLOW_ORIGIN, "*"),
-                    ];
-                    (StatusCode::OK, headers, file.contents()).into_response()
+                |ui| {
+                    ui.get_file(&path).map_or_else(
+                        || {
+                            (
+                                StatusCode::NOT_FOUND,
+                                ([(header::CONTENT_TYPE, "text/html")]),
+                                "<body><h1>404 Not Found</h1></body>",
+                            )
+                                .into_response()
+                        },
+                        |file| {
+                            // Serve included file
+                            let content_type = guess_content_type(file.path());
+                            let headers = [
+                                (header::CONTENT_TYPE, content_type),
+                                (header::ACCESS_CONTROL_ALLOW_ORIGIN, "*"),
+                            ];
+                            (StatusCode::OK, headers, file.contents()).into_response()
+                        },
+                    )
                 },
             )
-        },
-    )
+        }
+    }
```

Comment:
This function is getting a little messy, and with the `WebState` now possibly holding a bundled UI and possibly an override, we only ever need one of them. I suggest creating an enum for the `WebState` to control how the UI is served, and then implementing a function on that enum to get the actual response for it.

In my mind, it would look something like this:
```rust
#[derive(Clone)]
pub enum UISource {
    Bundled(&'static Dir<'static>),
    Dynamic(PathBuf),
    None,
}

impl UISource {
    pub fn get_ui(&self, path: &str) -> impl IntoResponse {
        match self {
            UISource::Bundled(dir) => Self::get_bundled_ui(dir, path).into_response(),
            UISource::Dynamic(dir) => Self::get_dynamic_ui(dir, path).into_response(),
            UISource::None => {
                (
                    StatusCode::NOT_FOUND,
                    ([(header::CONTENT_TYPE, "text/html")]),
                    "<body><h1>There is no UI bundled with this version of the application.</h1></body>",
                ).into_response()
            }
        }
    }

    fn get_bundled_ui(dir: &Dir<'static>, path: &str) -> impl IntoResponse {
        todo!("Old bundled UI code")
    }

    fn get_dynamic_ui(dir: &Path, path: &str) -> impl IntoResponse {
        todo!("New read from disk code")
    }
}

#[derive(Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct WebState {
    pub request: UnboundedSender<WebRequest>,
    pub ui: UISource,
}

impl WebState {
    #[must_use]
    pub fn new(ui: UISource) -> (Self, UnboundedReceiver<WebRequest>) {
        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
        (Self { request: tx, ui }, rx)
    }
}

// UI

async fn get_ui(
    State(state): State<WebState>,
    axum::extract::Path(path): axum::extract::Path<String>,
) -> impl IntoResponse {
    state.ui.get_ui(&path)
}
```